File: LeetCode_Java/twopoint/ContainerWithMostWater.java
package twopoint;

public class ContainerWithMostWater {

    public int maxArea(int[] height) {
        int l = 0;
        int r = height.length - 1;
        int maxArea = 0;

        while (l < r) {
            int h = Math.min(height[l], height[r]);
            maxArea = Math.max(maxArea, h * (r - l));
            if (height[l] <= height[r]) {
                l++;
            } else {
                r--;
            }
        }

        return maxArea;
    }

    public static void main(String[] args) {
        ContainerWithMostWater solution = new ContainerWithMostWater();

        // Test case
        int[] heights = {1, 8, 6, 2, 5, 4, 8, 3, 7};
        System.out.println("Maximum area: " + solution.maxArea(heights)); // Output: 49
    }
}

File: LeetCode_Python/twopoint/ContainerWithMostWater.py
class ContainerWithMostWater:
    def maxArea(self, height: list[int]) -> int:
        res = 0

        left, right = 0, len(height) - 1

        while left < right:
            lh, rh = height[left], height[right]
            res = max(res, min(lh, rh) * (right - left))
            if lh < rh:
                left += 1
            else:
                right -= 1

        return res


if __name__ == "__main__":
    solution = ContainerWithMostWater()

    height1 = [1, 8, 6, 2, 5, 4, 8, 3, 7]
    print("Expected Output: 49")
    print("Actual Output:", solution.maxArea(height1))

    height2 = [1, 1]
    print("Expected Output: 1")
    print("Actual Output:", solution.maxArea(height2))

    height3 = [4, 3, 2, 1, 4]
    print("Expected Output: 16")
    print("Actual Output:", solution.maxArea(height3))
